src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,246 @@
+      trainLabelsTemp.n_cols);
+  arma::Row<size_t> testLabels = arma::zeros<arma::Row<size_t>>(1,
+      testLabelsTemp.n_cols);
+
+  for (size_t i = 0; i < trainLabelsTemp.n_cols; ++i)
+    trainLabels(i) = arma::as_scalar(trainLabelsTemp.col(i));
+
+  for (size_t i = 0; i < testLabelsTemp.n_cols; ++i)
+    testLabels(i) = arma::as_scalar(testLabelsTemp.col(i));
+
+  for (size_t i = 0; i < trainData.n_cols; i++)
+  {
+    tempRadius = arma::norm(trainData.col(i));
+    if (radius < tempRadius)
+      radius = tempRadius;
+  }
+
+  size_t poolSize = 2;
+  radius *= 1.3;
+
+  arma::mat output;
+  arma::mat XRbm(hiddenLayerSize + poolSize * hiddenLayerSize,
+      trainData.n_cols);
+  arma::mat YRbm(hiddenLayerSize + poolSize * hiddenLayerSize,
+      testLabels.n_cols);
+
+  XRbm.zeros();
+  YRbm.zeros();
+  double slabPenalty = 8;
+
+  RBM<GaussianInitialization, arma::mat, SpikeSlabRBM> modelssRBM(trainData,
+      gaussian, trainData.n_rows, hiddenLayerSize, batchSize, 1, 1, poolSize,
+      slabPenalty, radius);
+
+  size_t numRBMIterations = trainData.n_cols * numEpoches;
+  numRBMIterations /= batchSize;
+
+  optimization::StandardSGD msgd(0.02, batchSize, numRBMIterations, 0, true);
+  modelssRBM.Reset();
+  modelssRBM.VisiblePenalty().fill(5);
+  modelssRBM.SpikeBias().fill(1);
+
+  modelssRBM.Train(msgd);
+  for (size_t i = 0; i < trainData.n_cols; i++)
+  {
+    modelssRBM.HiddenMean(std::move(trainData.col(i)),
+        std::move(output));
+    XRbm.col(i) = output;
+  }
+
+  for (size_t i = 0; i < testData.n_cols; i++)
+  {
+    modelssRBM.HiddenMean(std::move(testData.col(i)),
+      std::move(output));
+    YRbm.col(i) = output;
+  }
+  const size_t numClasses = 10; // Number of classes.
+  const size_t numBasis = 5; // Parameter required for L-BFGS algorithm.
+  const size_t numIterations = 100; // Maximum number of iterations.
+
+  L_BFGS ssRbmOptimizer(numBasis, numIterations);
+  SoftmaxRegression ssRbmRegressor(XRbm, trainLabels, numClasses,
+        0.001, false, ssRbmOptimizer);
+  double ssRbmClassificationAccuracy = ssRbmRegressor.ComputeAccuracy(
+      YRbm, testLabels);
+
+  BOOST_REQUIRE_GE(ssRbmClassificationAccuracy, 76.18);
+}
+
+template<typename MatType = arma::mat>
+void BuildVanillaNetwork(MatType& trainData,
+                         const size_t hiddenLayerSize)
+{
+  MatType output;
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, MatType, BinaryRBM> model(trainData, gaussian,
+      trainData.n_rows, hiddenLayerSize, 1, 1, 1, 2, 8, 1, true);
+
+  model.Reset();
+  // Set the parameters from a learned RBM Sklearn random state 23
+  model.Parameters() = MatType(
+      "-0.23224054, -0.23000632, -0.25701271, -0.25122418, -0.20716651,"
+      "-0.20962217, -0.59922456, -0.60003836, -0.6, -0.625, -0.475;");
+
+  // Check free energy
+  arma::Mat<float> freeEnergy = MatType(
+      "-0.87523715, 0.50615066, 0.46923476, 1.21509084;");
+  arma::vec calculatedFreeEnergy(4);
```

Comment:
we could use ``calculatedFreeEnergy(4, arma::fill::zeros);`` to avoid the ``zeros();`` call in the next line.